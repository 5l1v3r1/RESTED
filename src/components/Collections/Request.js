import React from 'react';
import { connect } from 'react-redux';
import { DragSource, DropTarget } from 'react-dnd';
import { ListGroup } from 'react-bootstrap';
import flow from 'lodash.flow';

import IconButton from 'components/IconButton';
import * as Actions from 'store/collections/actions';

import { StyledRequest, AsideButtons, MainContent } from './StyledComponents';
import * as Type from './dropTypes';

/**
 * Specifies the drag source contract.
 * Only `beginDrag` function is required.
 */
const requestSource = {
  canDrag() {
    // TODO Disallow drag if editing name
    return true;
  },

  beginDrag({ id, index, collectionIndex }) {
    return { id, index, collectionIndex };
  },
};

const requestTarget = {
  hover(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dragCollectionIndex = monitor.getItem().collectionIndex;
    const dragUUID = monitor.getItem().id;
    const hoverIndex = props.index;
    const hoverCollectionIndex = props.collectionIndex;
    const hoverUUID = props.id;

    // Abort further processing if dragged item is hovering over itself
    if (dragUUID === hoverUUID) {
      return;
    }

    // Note: we're mutating the monitor item here.
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    /* eslint-disable no-param-reassign */
    monitor.getItem().index = hoverIndex;
    monitor.getItem().collectionIndex = hoverCollectionIndex;
    /* eslint-enable no-param-reassign */

    // Update redux orders
    props.reorderRequest({
      // Source
      collectionIndex: dragCollectionIndex,
      requestIndex: dragIndex,
    }, {
      // Target
      collectionIndex: hoverCollectionIndex,
      requestIndex: hoverIndex,
    });
  },
};

function Request(props) {
  const {
    connectDragSource,
    connectDropTarget,
    isDragging,
    id,
    collectionIndex,
    method,
    url,
    deleteRequest,
  } = props;

  return connectDragSource(connectDropTarget(
    <div> {/* Need a wrapper div for React DnD support */}
      <StyledRequest isDragging={isDragging}>
        <ListGroup componentClass="div">
          <div className="list-group-item">
            <AsideButtons>
              <IconButton
                tooltip="Edit"
                icon="cog"
              />
              <IconButton
                tooltip="Delete"
                icon="trash"
                onClick={() => deleteRequest(id, collectionIndex)}
              />
            </AsideButtons>

            <MainContent>
              <h4>{method}</h4>
              {url}
            </MainContent>
          </div>
        </ListGroup>
      </StyledRequest>
    </div>,
  ));
}

export default flow(
  DropTarget(Type.Request, requestTarget, connector => ({
    connectDropTarget: connector.dropTarget(),
  })),
  DragSource(Type.Request, requestSource, (connector, monitor) => ({
    connectDragSource: connector.dragSource(),
    isDragging: monitor.isDragging(),
  })),
  connect(null, Actions),
)(Request);

