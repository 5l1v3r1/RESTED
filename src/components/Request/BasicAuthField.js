import React, { PropTypes } from 'react';
import { Col, FormGroup, FormControl } from 'react-bootstrap';

import Collapsable from '../Collapsable';

export function BasicAuthField(fields) {
  return (
    <Collapsable
      title="Basic auth"
      id="basicAuth"
    >
      <FormGroup controlId="method">
        <Col xs={5}>
          <FormControl
            type="text"
            placeholder="Username"
            {...fields.basicAuth.username.input}
          />
        </Col>
        <Col xs={5}>
          <FormControl
            type="text"
            placeholder="Password"
            {...fields.basicAuth.password.input}
          />
        </Col>
        <Col xs={2}>
          Show password?
        </Col>
      </FormGroup>
    </Collapsable>
  );
}

BasicAuthField.propTypes = {
  basicAuth: PropTypes.shape({
    username: PropTypes.shape({
      input: PropTypes.shape({}).isRequired,
    }).isRequired,
    password: PropTypes.shape({
      input: PropTypes.shape({}).isRequired,
    }).isRequired,
  }).isRequired,
};

export default BasicAuthField;

