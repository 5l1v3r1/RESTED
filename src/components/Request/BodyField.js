import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Field, FieldArray } from 'redux-form';
import { FormGroup, FormControl, Checkbox, Row, Col } from 'react-bootstrap';

import Fonticon from 'components/Fonticon';
import Collapsable from 'components/Collapsable';
import IconButton from 'components/IconButton';
import * as RequestActions from 'store/request/actions';

import { UnstyledButton } from './StyledComponents';

function renderField({ input, placeholder }) {
  return (
    <FormControl
      type="text"
      placeholder={placeholder}
      {...input}
    />
  );
}

renderField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  placeholder: PropTypes.string.isRequired,
};

function renderFormDataFields({ fields, meta, setUseFormData }) {
  return (
    <Collapsable
      title="Request body"
      id="requestBody"
    >
      <Col xs={12}>
        <Checkbox
          checked
          onChange={() => { setUseFormData(false); }}
        >
          Use form data
        </Checkbox>
      </Col>
      {fields.map((field, key) => (
        <FormGroup
          key={key}
          controlId={`requestBodyGroup${key}`}
          validationState={meta.invalid ? 'error' : null}
        >
          <Col xs={5}>
            <Field
              name={`${field}.name`}
              component={renderField}
              placeholder="Name"
            />
          </Col>
          <Col xs={6}>
            <Field
              name={`${field}.value`}
              component={renderField}
              placeholder="Value"
            />
          </Col>
          <Col xs={1}>
            <IconButton
              id={`removeFormDataButton${key}`}
              tooltip="Remove form field"
              icon="trash"
              onClick={() => fields.remove(key)}
            />
          </Col>
        </FormGroup>
      ))}
      <Row>
        <Col xs={12}>
          <UnstyledButton
            id="addParameter"
            bsStyle="link"
            onClick={() => fields.push({})}
          >
            <Fonticon icon="plus" />
            Add parameter
          </UnstyledButton>
        </Col>
      </Row>
    </Collapsable>
  );
}

/* eslint-disable react/no-unused-prop-types */
renderFormDataFields.propTypes = {
  setUseFormData: PropTypes.func.isRequired,
  fields: PropTypes.shape({
    map: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
  meta: PropTypes.shape({
    invalid: PropTypes.bool.isRequired,
  }).isRequired,
};
/* eslint-enable react/no-unused-prop-types */

function renderDataField({ input, setUseFormData }) {
  return (
    <Collapsable
      title="Request body"
      id="requestBody"
    >
      <Col xs={12}>
        <Checkbox
          onChange={() => { setUseFormData(true); }}
        >
          Use form data
        </Checkbox>
      </Col>

      <Col xs={12}>
        <FormGroup controlId="requestBody">
          <FormControl
            componentClass="textarea"
            rows={10}
            {...input}
          />
        </FormGroup>
      </Col>
    </Collapsable>
  );
}

renderDataField.propTypes = {
  input: PropTypes.shape({}).isRequired,
  setUseFormData: PropTypes.func.isRequired,
};

export function BodyField({ useFormData, setUseFormData }) {
  if (useFormData) {
    return (
      <FieldArray
        name="formData"
        component={renderFormDataFields}
        setUseFormData={setUseFormData}
      />
    );
  }

  return (
    <Field
      name="data"
      component={renderDataField}
      setUseFormData={setUseFormData}
    />
  );
}

BodyField.propTypes = {
  useFormData: PropTypes.bool,
  setUseFormData: PropTypes.func.isRequired,
};

const mapStateToProps = ({ request: { useFormData } }) => ({
  useFormData,
});

export default connect(mapStateToProps, RequestActions)(BodyField);

