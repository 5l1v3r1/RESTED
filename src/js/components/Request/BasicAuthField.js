import React, { PropTypes } from 'react';
import { FormGroup } from 'react-bootstrap';

import Collapsable from '../Collapsable';

export function BasicAuthField({ meta }) {
  return (
    <Collapsable
      title="Basic auth"
      id="basicAuth"
    >
      <FormGroup
        controlId="method"
        validationState={meta.invalid ? 'error' : undefined}
      >
        BasicAuth
      </FormGroup>
    </Collapsable>
  );
}

BasicAuthField.propTypes = {
  /* eslint-disable react/forbid-prop-types */
  meta: PropTypes.object.isRequired,
  /* eslint-enable react/forbid-prop-types */
};

export default BasicAuthField;

