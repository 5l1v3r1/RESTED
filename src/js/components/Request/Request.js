import React from 'react';
import { Col } from 'react-bootstrap';

export default function Request() {
  return (
    <Col xs={12} sm={8}>
      <h2>
        Request
      </h2>
    </Col>
  );
}

