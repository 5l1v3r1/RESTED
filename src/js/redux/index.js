import { combineReducers } from 'redux';
import { reducer as form } from 'redux-form';

import request from './request/reducer';

export default combineReducers({
  request,
  form,
});

