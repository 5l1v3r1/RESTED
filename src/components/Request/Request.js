import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel } from 'react-bootstrap';

import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

import * as Actions from '../../store/request/actions';
import { DEFAULT_REQUEST } from '../../constants/constants';

function Titlebar() {
  return (
    <h2>
      Request
    </h2>
  );
}

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <form onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: 'request',
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

// TODO this
/* eslint-disable no-func-assign */
Request = reduxForm(formOptions)(Request);
Request = connect(mapStateToProps, Actions)(Request);
/* eslint-enable no-func-assign */
export default Request;

