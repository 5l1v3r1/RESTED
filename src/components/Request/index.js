import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, Fields, FieldArray, propTypes } from 'redux-form';
import { Panel, Form } from 'react-bootstrap';
import flow from 'lodash.flow';

import * as requestActions from 'store/request/actions';
import requestValidation from 'utils/requestValidation';
import { DEFAULT_REQUEST } from 'constants/constants';

import Titlebar from './Titlebar';
import URLField from './URLField';
import MethodField from './MethodField';
import HeadersField from './HeadersField';
import BasicAuthField from './BasicAuthField';
import BodyField from './BodyField';

export const requestForm = 'request';

function Request({ placeholderUrl, handleSubmit, sendRequest }) {
  return (
    <Panel header={<Titlebar />}>
      <Form horizontal onSubmit={handleSubmit(sendRequest)}>
        <Field
          name="url"
          component={URLField}
          placeholderUrl={placeholderUrl}
        />
        <Field
          name="method"
          component={MethodField}
        />
        <FieldArray
          name="headers"
          component={HeadersField}
        />
        <Fields
          names={['basicAuth.username', 'basicAuth.password']}
          component={BasicAuthField}
        />
        <BodyField />
      </Form>
    </Panel>
  );
}

Request.propTypes = {
  placeholderUrl: PropTypes.string,
  ...propTypes,
};

const formOptions = {
  form: requestForm,
  validate: requestValidation,
};

const mapStateToProps = ({ request: { useFormData, placeholderUrl } }) => ({
  useFormData,
  placeholderUrl,
  initialValues: DEFAULT_REQUEST,
});

export { Request };

export default flow(
  reduxForm(formOptions),
  connect(mapStateToProps, requestActions),
)(Request);

