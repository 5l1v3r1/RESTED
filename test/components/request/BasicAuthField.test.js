import React from 'react';
import renderer from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';
import { Provider, connect } from 'react-redux';

/* eslint-disable import/no-unresolved */
import { BasicAuthField } from 'components/Request/BasicAuthField';
import collapsable from 'store/collapsable/reducer';

const makeStore = (initial = {}) => createStore(
  combineReducers({ collapsable }),
  { collapsable: initial }
);

describe('BasicAuthField', () => {
  const store = makeStore({
    collapsable: {},
  });

  it('should match the previous snapshot when valid', () => {
    // Mock props
    BasicAuthField.defaultProps = {
      input: {},
      meta: {
        invalid: false,
        valid: true,
      },
    };

    const Decorated = connect()(BasicAuthField);

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match the previous snapshot when invalid', () => {
    // Mock props
    BasicAuthField.defaultProps = {
      input: {},
      meta: {
        invalid: true,
        valid: false,
      },
    };

    const Decorated = connect()(BasicAuthField);

    const tree = renderer.create(
      <Provider store={store}>
        <Decorated />
      </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

