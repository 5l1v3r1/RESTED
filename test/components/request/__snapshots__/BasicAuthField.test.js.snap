exports[`BasicAuthField should match the previous snapshot when invalid 1`] = `
<div
  className="row">
  <button
    className="btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button">
    <h3>
      Basic auth
      <i
        className="fa fa-angle-right" />
    </h3>
  </button>
  <div
    aria-expanded={null}
    className="collapse col-xs-12">
    <div
      className="form-group has-error">
      BasicAuth
    </div>
  </div>
</div>
`;

exports[`BasicAuthField should match the previous snapshot when valid 1`] = `
<div
  className="row">
  <button
    className="btn btn-link"
    disabled={false}
    onClick={[Function]}
    type="button">
    <h3>
      Basic auth
      <i
        className="fa fa-angle-right" />
    </h3>
  </button>
  <div
    aria-expanded={null}
    className="collapse col-xs-12">
    <div
      className="form-group">
      BasicAuth
    </div>
  </div>
</div>
`;
